File: backend/app/schemas/session_assistant.py
"""Schemas transitorios para el Asistente IA de sesiones.

Ningún campo de estos schemas toca la DB; todo es stateless
request/response. Los athlete_ids que llegan en el request se
descartan en el servidor después de calcular los conteos agregados
(nunca se reenvían al LLM ni se retornan al cliente).
"""

from __future__ import annotations

import enum
from datetime import date, time

from pydantic import BaseModel, Field, field_validator, model_validator


# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------


class AthleteCallUpCriterion(str, enum.Enum):
    """Propuesta no-identificante del grupo convocado.

    El frontend resuelve este criterio contra el roster local para
    llenar ``convocados_athlete_ids``. Nunca se transmiten ids o nombres.
    """

    todos_convocados = "todos_convocados"
    grupo_10_12 = "grupo_10_12"
    grupo_13_15 = "grupo_13_15"
    ninguno = "ninguno"


# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------


class SessionClarifyRequest(BaseModel):
    """Cuerpo del POST /clarify.

    ``selected_athlete_ids`` se usa únicamente en el servidor para
    calcular el age_mix agregado. Nunca se envía al LLM ni se registra
    en logs.
    """

    intent_text: str | None = Field(
        default=None,
        max_length=500,
        description="Descripción libre de la sesión (cualquier idioma).",
    )
    selected_athlete_ids: list[int] = Field(
        default_factory=list,
        description="IDs de atletas seleccionados (solo para calcular age_mix).",
    )


class SessionAnswer(BaseModel):
    """Una respuesta a una pregunta de clarificación retornada."""

    question_id: str = Field(
        ...,
        description="Coincide con el ``id`` de la pregunta retornada.",
    )
    selected_labels: list[str] = Field(
        default_factory=list,
        description="Etiquetas de opciones elegidas.",
    )
    other_text: str | None = Field(
        default=None,
        max_length=300,
        description="Texto libre cuando se eligió 'Otro'.",
    )


class SessionDraftRequest(BaseModel):
    """Cuerpo del POST /draft."""

    intent_text: str | None = Field(
        default=None,
        max_length=500,
    )
    selected_athlete_ids: list[int] = Field(
        default_factory=list,
    )
    answers: list[SessionAnswer] = Field(
        default_factory=list,
        description="Respuestas del coach a las preguntas de clarificación.",
    )


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------


class ClarifyOption(BaseModel):
    """Una opción de respuesta dentro de una pregunta de clarificación."""

    label: str = Field(..., min_length=1, max_length=40)
    description: str = Field(..., min_length=1, max_length=120)


class ClarifyQuestion(BaseModel):
    """Una pregunta de clarificación generada por la IA."""

    id: str = Field(..., description="Identificador estable, ej. 'q1'.")
    header: str = Field(..., min_length=1, max_length=12)
    question: str = Field(..., min_length=1, max_length=160)
    multi_select: bool = False
    allow_other: bool = False
    options: list[ClarifyOption] = Field(
        ...,
        description="2–4 opciones.",
    )

    @field_validator("options")
    @classmethod
    def _validate_options_count(cls, v: list[ClarifyOption]) -> list[ClarifyOption]:
        if len(v) < 2 or len(v) > 4:
            raise ValueError(
                f"Cada pregunta debe tener entre 2 y 4 opciones; se recibieron {len(v)}."
            )
        return v


class SessionClarifyResponse(BaseModel):
    """Respuesta del endpoint /clarify."""

    questions: list[ClarifyQuestion] = Field(
        default_factory=list,
        description="0–4 preguntas; 0 significa que el coach puede pasar directo a /draft.",
    )
    model: str = Field(..., description="ID del modelo utilizado.")

    @field_validator("questions")
    @classmethod
    def _validate_questions_count(
        cls, v: list[ClarifyQuestion]
    ) -> list[ClarifyQuestion]:
        if len(v) > 4:
            raise ValueError(
                f"La respuesta no puede contener más de 4 preguntas; se recibieron {len(v)}."
            )
        return v


class SessionDraftResponse(BaseModel):
    """Respuesta del endpoint /draft — mapea campo a campo con el wizard."""

    technical_focus: str = Field(..., min_length=1, max_length=200)
    objectives: str | None = Field(default=None, max_length=1000)
    description: str | None = Field(default=None, max_length=2000)
    duration_min: int = Field(..., ge=15, le=240)
    session_kind: str = Field(
        default="entrenamiento",
        description="Valor de SessionKind.",
    )
    location: str | None = Field(default=None, max_length=200)
    scheduled_date: date | None = None
    scheduled_start_time: time | None = None
    athlete_call_up: AthleteCallUpCriterion = AthleteCallUpCriterion.ninguno
    notes: str | None = Field(default=None, max_length=500)
    model: str = Field(..., description="ID del modelo utilizado.")

    @field_validator("session_kind")
    @classmethod
    def _validate_session_kind(cls, v: str) -> str:
        from app.models.training_session import SessionKind

        valid = {e.value for e in SessionKind}
        if v not in valid:
            raise ValueError(
                f"session_kind '{v}' inválido. Valores aceptados: {sorted(valid)}."
            )
        return v

File: frontend/src/api/sessionAssistant.ts
/**
 * API client para el módulo Asistente IA de sesiones (Feature 006).
 *
 * Endpoints:
 *   POST /api/clubs/{clubId}/session-assistant/clarify
 *   POST /api/clubs/{clubId}/session-assistant/draft
 *
 * Privacy contract: selected_athlete_ids se usa solo server-side para calcular
 * mezcla de edades; nunca se envía nombre o identificador de menor al modelo IA.
 */
import { apiClient } from "@/api/client";

// ---------------------------------------------------------------------------
// Tipos — espejo del contrato en contracts/session-assistant.md
// ---------------------------------------------------------------------------

export type AthleteCallUpCriterion =
  | "todos_convocados"
  | "grupo_10_12"
  | "grupo_13_15"
  | "ninguno";

export interface ClarifyOption {
  label: string;
  description: string;
}

export interface ClarifyQuestion {
  id: string;
  header: string;
  question: string;
  multi_select: boolean;
  allow_other: boolean;
  options: ClarifyOption[];
}

export interface SessionClarifyResponse {
  questions: ClarifyQuestion[];
  model: string;
}

export interface SessionAnswer {
  question_id: string;
  selected_labels: string[];
  other_text: string | null;
}

export interface SessionDraftResponse {
  technical_focus: string;
  objectives: string | null;
  description: string | null;
  duration_min: number;
  session_kind: "entrenamiento" | "actividad_conjunta" | "salida" | "otro";
  location: string | null;
  scheduled_date: string | null;
  scheduled_start_time: string | null;
  athlete_call_up: AthleteCallUpCriterion;
  notes: string | null;
  model: string;
}

// ---------------------------------------------------------------------------
// Request shapes
// ---------------------------------------------------------------------------

export interface SessionClarifyRequest {
  intent_text?: string | null;
  selected_athlete_ids?: number[];
}

export interface SessionDraftRequest {
  intent_text?: string | null;
  selected_athlete_ids?: number[];
  answers?: SessionAnswer[];
}

// ---------------------------------------------------------------------------
// Funciones de cliente
// ---------------------------------------------------------------------------

const BASE = (clubId: number) =>
  `/api/clubs/${clubId}/session-assistant`;

/**
 * Solicita un lote de preguntas de clarificación al asistente IA.
 * 0 preguntas en la respuesta → el cliente puede llamar directamente a draft().
 */
export async function clarify(
  clubId: number,
  payload: SessionClarifyRequest,
): Promise<SessionClarifyResponse> {
  const response = await apiClient.post<SessionClarifyResponse>(
    `${BASE(clubId)}/clarify`,
    payload,
  );
  return response.data;
}

/**
 * Solicita un borrador de sesión basado en la intención y las respuestas.
 * Acepta respuestas parciales o vacías (FR-015).
 */
export async function draft(
  clubId: number,
  payload: SessionDraftRequest,
): Promise<SessionDraftResponse> {
  const response = await apiClient.post<SessionDraftResponse>(
    `${BASE(clubId)}/draft`,
    payload,
  );
  return response.data;
}
